Deprecate Mailbox v1 and support v2

aquartier opened the pull request:
_On June 6th, 2019, support for legacy Mailbox API 1.0 will be shut down. To ensure that all existing integrations continue to function properly and are compatible with a supported API, GirlScout must be moved to Mailbox API 2.0._

**TESTING:**

Just check out this branch and run `rake test`

**NEW:**
- [x] Token model
- [x] Listing API for Customer, Mailbox, Users and Conversation classes.
- [x] Error message parser for API v2.

**BREAKING CHANGES:**
- [x] Kill the API v1.
- [x] Resouce listing and retrieving.
- [x] Nested resources.
- [x] ~Add some hacking stuff for Mailbox and Conversation APIs.~
- [x] Deprecated the Search API.

**MISC:**
- [x] Update test fixtures.
- [x] Add default configurations.

chakrit commented on lib/girlscout/conversation.rb:
```diff
@@ -6,35 +6,58 @@ class Conversation < GirlScout::Object
         attributes = attributes.as_json if attributes.respond_to?(:as_json)
-        attributes['reload'] ||= true
-        Conversation.new(resource.post(payload: attributes)['item'])
+        find(resource.post(payload: attributes))
```
Although this is nice, it deviates from the inteded semantics of the HelpScout API. This means, users can't rely on reading the HelpScout API to figure out what our library does (hidden magic.)

API bindings shouldn't add too many re-interpretation of things otherwise it gets pretty confusing because instead of understanding one set of API, now you need to understand two. Then the girlscout gem is no longer transparent to the user.

I think it's simpler here to just return a representation of 201 or whatever to the consumer (so we're as close to HelpScout API semantics as possible, rather than making this into 2 http calls to preserve old v1 semantics.)

We can still provide a quick explicit helper method to turn that back into a proper objects.

aquartier replied:
Good point. Previously, I want to preserve the current behavior library (v1) but I forget that we must follow the new API behaviors and not deviate them : )

aquartier replied:
With the above reason, ☝️ I also remove the Search model since it no longer exists in API v2.

aquartier replied:
And added some of `.list` (with query) endpoints to the models to make them more reflect with the API documents.

chakrit commented on lib/girlscout/conversation.rb:
```diff
@@ -6,35 +6,58 @@ class Conversation < GirlScout::Object
+        if key?('customer')
+          Customer.new(self['customer'] || {})
+        else
+          Customer.new(self['primary_customer'] || {})
```
Chain-able here.

```rb
Customer.new(self['customer'] || self['primary_customer'] || {})
```

chakrit commented on lib/girlscout/conversation.rb:
```diff
@@ -6,35 +6,58 @@ class Conversation < GirlScout::Object
 
     def mailbox
-      @mailbox ||= Mailbox.new(self['mailbox'] || {})
+      @threads ||= begin
```
`def mailbox` initializing a `@threads` .. bug?

aquartier replied:
Yeah, my mistake from copy-paste.

chakrit commented on lib/girlscout/resource.rb:
```diff
@@ -6,40 +6,49 @@ module GirlScout
+      when 201
+        response.headers['Resource-ID']
+      else
+        raise GirlScout::Error, message: error_message(response.body), code: response.status
```
Another indirection here. `message -> error_message`, I think better to just rename attributes on `GirlScout::Error` to match.

aquartier replied:
An `error_message` is just a parser method for the Resource class. I would like to keep the attr name as the same otherwise we need to call `GirScout::Error.error_message` to get the message which seems redundant.

chakrit commented on lib/girlscout/token.rb:
```diff
@@ -0,0 +1,50 @@
+
+      private
+
+      def request
```
How about `request_new` to be more clear.

aquartier replied:
Token model has been rewritten. I renamed to `refresh` method. Please review again.

chakrit commented on lib/girlscout/token.rb:
```diff
@@ -0,0 +1,50 @@
+
+        response = JSON.parse(response.body)
+        @token = response['access_token']
+        @expires_at = Time.now + response['expires_in']
```
Alternatively, we could also `include HasAttributes` into this thing and get a nicer accessor on `access_token` and `expires_in` ?

(also, yes, name indirection again here. probably better to stick to helpscout's name)

aquartier commented on lib/girlscout/resource.rb:
```diff
@@ -6,40 +6,49 @@ module GirlScout
+          'Authorization' => "Bearer #{Token.retrieve}"
+        }
+        options[:body] = JSON.generate(payload) if payload
+        options[:query] = URI.encode_www_form(query) if query
```
@chakrit I tried before but sadly  will break the `vcr` request. https://github.com/vcr/vcr/issues/428

aquartier commented on Gemfile.lock:
```diff
@@ -93,7 +92,8 @@ DEPENDENCIES
   pry (~> 0.11)
   rake (~> 12.3)
   rubocop (~> 0.58)
+  timecop (~> 0.9)
```
agreed and seems overkill, replacing with simple stub now : )

aquartier commented:
@chakrit changes after code review:
- Refactored the Token model by rename to AccessToken, stick to HelpScout's attr name and made it stateless and control by the consumer (Resource class).
- Added listing API for Customer, Mailbox, Users and Conversation classes.
- Removed Search API since it was no longer available in API v2.
- Use stub to control Time instead of the `timecop` gem.
- Updated README with more usage example.
